GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,119 @@
+﻿using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace GVFS.Platform.Mac
+{
+    public class MacFileBasedLock : FileBasedLock
+    {
+        private const int InvalidFileDescriptor = -1;
+
+        private const int LockSh = 1; // #define LOCK_SH   1    /* shared lock */
+        private const int LockEx = 2; // #define LOCK_EX   2    /* exclusive lock */
+        private const int LockNb = 4; // #define LOCK_NB   4    /* don't block when locking */
+        private const int LockUn = 8; // #define LOCK_UN   8    /* unlock */
+
+        private int lockFileDescriptor;
+        private bool lockAcquired;
+
+        public MacFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath,
+            string signature)
+            : base(fileSystem, tracer, lockPath, signature) 
+        {
+            this.lockFileDescriptor = InvalidFileDescriptor;
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockAcquired)
+            {
+                return true;
+            }
+
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = Creat(this.LockPath, Convert.ToInt32("644", 8));
```

Comment:
Is 644 always the right mode? What if someone has left this file behind with a different mode on it?

Reply:
I think this code will be clearer with the switch to `open` (`creat` is the same as `open` with the flags we're interested in here).  

The `mode` parameter is only used when creating a new file, it's ignored if the file already exists:

> The mode argument specifies the file mode bits be applied when
              a new file is created.  This argument must be supplied when
              O_CREAT or O_TMPFILE is specified in flags; if neither O_CREAT
              nor O_TMPFILE is specified, then mode is ignored.  The
              effective mode is modified by the process's umask in the usual
              way: in the absence of a default ACL, the mode of the created
              file is (mode & ~umask).  Note that this mode applies only to
              future accesses of the newly created file;

http://man7.org/linux/man-pages/man2/open.2.html